File: src/KnowledgeOps.Api/Controllers/Models/DashboardModels.cs
namespace KnowledgeOps.Api.Controllers.Models;

// ── Shared types ──────────────────────────────────────────────────────────────

public sealed record DashboardPeriodResponse(
    DateTimeOffset From,
    DateTimeOffset To);

public sealed record DashboardCostResponse(
    bool Available,
    decimal? EstimatedTotal);

public sealed record DashboardTokensResponse(
    long? Input,
    long? Output,
    long? Total);

// ── Overview response ─────────────────────────────────────────────────────────

public sealed record DashboardOverviewResponse(
    DashboardPeriodResponse Period,
    int QuestionsAsked,
    int ActiveUsers,
    int DocumentsUploaded,
    int DocumentsProcessed,
    int DocumentsFailed,
    long? AverageResponseLatencyMs,
    int InsufficientContextCount,
    int ProviderFailureCount,
    int UsefulFeedbackCount,
    int NotUsefulFeedbackCount,
    DashboardCostResponse Cost);

// ── Documents response ────────────────────────────────────────────────────────

public sealed record DashboardDocumentsResponse(
    DashboardPeriodResponse Period,
    int Uploaded,
    int Processing,
    int Processed,
    int Failed,
    int RetrievalDisabled);

// ── Chat response ─────────────────────────────────────────────────────────────

public sealed record DashboardChatResponse(
    DashboardPeriodResponse Period,
    int QuestionsAsked,
    int ActiveUsers,
    long? AverageResponseLatencyMs,
    long? RetrievalLatencyMs,
    long? GenerationLatencyMs,
    long? TotalRagLatencyMs,
    int InsufficientContextCount,
    int ProviderFailureCount,
    DashboardTokensResponse Tokens,
    DashboardCostResponse Cost);

// ── Feedback response ─────────────────────────────────────────────────────────

public sealed record DashboardFeedbackResponse(
    DashboardPeriodResponse Period,
    int Useful,
    int NotUseful,
    int Total);

File: frontend/src/app/features/dashboard/models/dashboard.models.ts
// Dashboard API response models.
// All nullable fields are typed as X | null (not undefined).

export interface DashboardPeriod {
  from: string;
  to: string;
}

export interface DashboardCostInfo {
  available: boolean;
  // NEVER show $0.00 when available === false. Show "Not available" instead.
  estimatedTotal: number | null;
}

export interface DashboardTokenInfo {
  input: number | null;
  output: number | null;
  total: number | null;
}

export interface DashboardOverviewResponse {
  period: DashboardPeriod;
  questionsAsked: number;
  activeUsers: number;
  documentsUploaded: number;
  documentsProcessed: number;
  documentsFailed: number;
  // null means unavailable — never show 0 for null
  averageResponseLatencyMs: number | null;
  insufficientContextCount: number;
  providerFailureCount: number;
  usefulFeedbackCount: number;
  notUsefulFeedbackCount: number;
  cost: DashboardCostInfo;
}

export interface DashboardDocumentsResponse {
  period: DashboardPeriod;
  uploaded: number;
  processing: number;
  processed: number;
  failed: number;
  retrievalDisabled: number;
}

export interface DashboardChatResponse {
  period: DashboardPeriod;
  questionsAsked: number;
  activeUsers: number;
  averageResponseLatencyMs: number | null;
  retrievalLatencyMs: number | null;
  generationLatencyMs: number | null;
  totalRagLatencyMs: number | null;
  insufficientContextCount: number;
  providerFailureCount: number;
  tokens: DashboardTokenInfo;
  cost: DashboardCostInfo;
}

export interface DashboardFeedbackResponse {
  period: DashboardPeriod;
  useful: number;
  notUseful: number;
  total: number;
}

export interface DashboardDateParams {
  from?: string;
  to?: string;
}
